src/coord/src/coord.rs

```diff
@@ -2206,31 +2206,32 @@ impl Coordinator {
         &mut self,
         tx: ClientTransmitter<ExecuteResponse>,
         mut session: Session,
-        action: EndTransactionAction,
+        mut action: EndTransactionAction,
     ) {
         let was_implicit = matches!(
             session.transaction(),
             TransactionStatus::InTransactionImplicit(_)
         );
+        // If the transaction has failed, we can only rollback.
+        if let (EndTransactionAction::Commit, TransactionStatus::Failed(_)) =
+            (&action, session.transaction())
+        {
+            action = EndTransactionAction::Rollback;
+        }
+        let response = ExecuteResponse::TransactionExited {
+            tag: action.tag(),
+            was_implicit,
+        };
         let rx = self.sequence_end_transaction_inner(&mut session, &action);
         match rx {
             Ok(Some(rx)) => {
                 tokio::spawn(async move {
-                    let tag = match rx.await {
-                        Ok(_) => action.tag(),
-                        // TODO: Try to return this error somehow.
-                        Err(_err) => EndTransactionAction::Rollback.tag(),
-                    };
-                    let result = Ok(ExecuteResponse::TransactionExited { tag, was_implicit });
+                    let result = rx.await.map(|_| response);
```

Comment:
I'm not really clear on why we are awaiting this channel given that we replace whatever we get with our own pre-determined response. I think a comment would help here?

Reply:
The output is a `Result<(), CoordError>`, so the `.map` only remaps a `Ok(())` result to `response`, but passes through any Err. Notice the lack of `?`, we aren't ignoring any errors here. Added a comment describing this a bit.